data_transfer/devices/dreem.py

```diff
@@ -46,41 +62,116 @@ def download_metadata(self) -> None:
+                    device_record = dreem_devices[0]
+                # Edge-case: multiple dreem headbands used, e.g., if one broke.
+                elif len(dreem_devices) > 1:
+                    # Determine usage based on weartime as it exists in UCAM
+                    device_record = ucam.record_by_wear_period_in_list(
+                        dreem_devices, recording.start, recording.end
+                    )
+                # Edge-case: device not logged with patient in UCAM
+                else:
+                    print(f"Metadata cannot be determined for:\n    {recording}")
+                    continue
+            else:
+                unknown += 1
+                print(f"Metadata cannot be determined for:\n    {recording}")
+                continue
 
-            start_wear = utils.format_weartime(history["checkout"], "inventory")
-            # NOTE: if device not returned the current day is used.
-            end_wear = utils.format_weartime(history["checkin"], "inventory")
             record = Record(
-                # NOTE: id is a unique uuid used to GET raw data from Dreem
-                filename=item["id"],
-                device_type=utils.DeviceType.DRM.name,
-                device_id=device_id,
+                filename=recording.id,
+                device_type=dtype,
                 patient_id=patient_id,
-                start_wear=start_wear,
-                end_wear=end_wear,
+                **asdict(device_record),
             )
 
             create_record(record)
+            print(f"Record Created:\n   {record}")
 
             path = Path(config.storage_vol / f"{record.filename}-meta.json")
             # Store metadata from memory to file
             utils.write_json(path, item)
 
+            print(f"Metadata saved to: {path}\n")
+        print(f"{known} records created and {unknown} NOT this session.")
+
+    def __recording_metadata(self, recording: dict) -> DreemRecording:
+        """
+        Maps data from Dreem response to class to simplify access/logging.
+        """
+        id = recording["id"]
+        device_id = recording["device"]
+        user_id = recording["user"]
+        # When the recording took place
+        start = datetime.fromtimestamp(recording["report"]["start_time"])
+        end = datetime.fromtimestamp(recording["report"]["stop_time"])
+
+        return DreemRecording(*(id, device_id, user_id, start, end))
+
+    def __patient_id_from_ucam(
+        self, device_serial: str, dreem_start: datetime, dreem_end: datetime
+    ) -> Optional[str]:
+        """
+        Determine PatientID by wear period of device in UCAM.
+        """
+        # NOTE/TODO: given this is a 1-1 mapping, why not use a local CSV?
+        device_id = inventory.device_id_by_serial(device_serial)
+        record = ucam.record_by_wear_period(device_id, dreem_start, dreem_end)
+        # TODO: inventory has small rate limit.
+        time.sleep(2.5)
```

Comment:
There is an API rate limit on the inventory of [150 requests by default](https://snipe-it.readme.io/reference#api-throttling). Critically, once the rate limit is met calls to the inventory's database appear to be disabled for a set period (about a minute) that would impact anyone using the inventory website.

We should increase this to a more reasonable limit, e.g., 3x current limit (450)? This requires modifying the source code within the docker container as there appears to yet be no way to do this via an `.env`.

Reply:
I got a connection error at metadata file nr 28 already - so quite a pressing issue. I am wondering if there are alternatives. Shot in the dark, but how about querying our past records (in the MongoDB) for some serials (if they are in there)? Should promptly return a result in recurring Dreem devices - which I expect is a common thing.

Reply:
Interesting ... not sure if the connection error is related (might be MongoDB?), but either way is a pressing issue.

> Shot in the dark, but how about querying our past records (in the MongoDB) for some serials (if they are in there)? Should promptly return a result in recurring Dreem devices - which I expect is a common thing.

We do not store serials in mongoDB although we might for convience to resolve this issue? Alternatives could be: (1) store serial->ID mapping in a CSV; (2) cache the API request; (3) store serials in MongoDB and do a lookup there.

Perhaps (3) is reasonable to store under the "meta" field?

Reply:
I've referenced this discussion on the issue one [the stack](https://github.com/ideafast/stack/issues/9#issuecomment-781258338)